Add StrokeLineWithColorAndThickness message to RemoteDisplayListRecorder

mattwoodrow opened the pull request:
#### 0eab85d1cfbec52489bb70f7bf2ad507c27d0fd1
<pre>
Add StrokeLineWithColorAndThickness message to RemoteDisplayListRecorder
<a href="https://bugs.webkit.org/show_bug.cgi?id=242772">https://bugs.webkit.org/show_bug.cgi?id=242772</a>

Reviewed by Simon Fraser.

There&apos;s some overhead per-message sent, and stroking lines is frequently combined with
updating the current stroke color and line thickness. Adding a special combination message
for this common case helps improve performance.

This also reduces the batch size, since we&apos;re now sending fewer (but larger) messages, and
having the batch size be too large results in unnecessary idle time on the receiving end.

This solution won&apos;t scale well, but hopefully is sufficient to handle the most common cases
in the short term.

I think an ideal solution would be to dynamically include state changes as part of the message
payload for all draw calls, rather than defining a specific new message for every possible
combination.

* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::strokePath):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordStrokeLineWithColorAndThickness):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::strokeLineWithColorAndThickness):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeLineWithColorAndThickness):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Canonical link: <a href="https://commits.webkit.org/252522@main">https://commits.webkit.org/252522@main</a>
</pre>

weinig commented on Source/WebKit/Platform/IPC/StreamClientConnection.h:
```diff
@@ -157,7 +157,7 @@ class StreamClientConnection final {
     std::optional<Semaphores> m_semaphores;
-    unsigned m_maxBatchSize { 40 }; // Number of messages marked as StreamBatched to accumulate before notifying the server.
+    unsigned m_maxBatchSize { 20 }; // Number of messages marked as StreamBatched to accumulate before notifying the server.
     unsigned m_batchSize { 0 };
```
Was this intended as part of the change? The ChangeLog didn't mention it and it seems a bit unrelated.

mattwoodrow replied:
It is! You're right that the ChangeLog definitely should mention it though, I'll do that now.

mattwoodrow committed:
Add StrokeLineWithColorAndThickness message to RemoteDisplayListRecorder
https://bugs.webkit.org/show_bug.cgi?id=242772

Reviewed by Simon Fraser.

There's some overhead per-message sent, and stroking lines is frequently combined with
updating the current stroke color and line thickness. Adding a special combination message
for this common case helps improve performance.

This also reduces the batch size, since we're now sending fewer (but larger) messages, and
having the batch size be too large results in unnecessary idle time on the receiving end.

This solution won't scale well, but hopefully is sufficient to handle the most common cases
in the short term.

I think an ideal solution would be to dynamically include state changes as part of the message
payload for all draw calls, rather than defining a specific new message for every possible
combination.

* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::strokePath):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordStrokeLineWithColorAndThickness):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::strokeLineWithColorAndThickness):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeLineWithColorAndThickness):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Canonical link: https://commits.webkit.org/252522@main

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
```diff
@@ -469,8 +469,15 @@ void Recorder::strokeRect(const FloatRect& rect, float lineWidth)
+        return;
+    }
+
+    appendStateChangeItemIfNecessary();
```
This call could be put before 

```
recordStrokeLine(path.inlineData<LineData>());
```

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
```diff
@@ -482,6 +489,8 @@ void Recorder::strokePath(const Path& path)
         return;
     }
+#else
+    appendStateChangeItemIfNecessary();
```
And this one could be moved just before 

```
recordStrokePath(path);
```

So it is clearer that every time we record a drawing command, we are sure the state was already updated.
